crates/turbopack-node/src/render/render_proxy.rs

```diff
@@ -44,105 +52,55 @@ pub async fn render_proxy(
+        path,
     )
-    .await
-    {
-        Ok(proxy_result) => Ok(proxy_result.cell()),
-        Err(err) => Ok(proxy_error(path, err, Some(operation)).await?),
-    }
-}
-
-async fn run_proxy_operation(
-    operation: &mut NodeJsOperation,
-    data: RenderDataVc,
-    body: BodyVc,
-    intermediate_asset: AssetVc,
-    intermediate_output_path: FileSystemPathVc,
-    project_dir: FileSystemPathVc,
-) -> Result<ProxyResult> {
-    let data = data.await?;
-    // First, send the render data.
-    operation
-        .send(RenderProxyOutgoingMessage::Headers { data: &data })
-        .await?;
-
-    let mut body = body.await?.read();
-    // Then, send the binary body in chunks.
-    while let Some(data) = body.next().await {
-        operation
-            .send(RenderProxyOutgoingMessage::BodyChunk { data: &data? })
-            .await?;
-    }
-
-    operation.send(RenderProxyOutgoingMessage::BodyEnd).await?;
+    .await?;
 
-    let (status, headers) = match operation.recv().await? {
-        RenderProxyIncomingMessage::Headers {
-            data: ResponseHeaders { status, headers },
-        } => (status, headers),
-        RenderProxyIncomingMessage::Error(error) => {
-            bail!(
-                trace_stack(
-                    error,
-                    intermediate_asset,
-                    intermediate_output_path,
-                    project_dir
-                )
-                .await?
-            )
-        }
+    let mut stream = render.read();
+    let first = match stream.next().await {
+        Some(Ok(f)) => f,
         _ => {
-            bail!("unexpected response from the Node.js process while reading response headers")
+            // If an Error was received first, then it would have been
+            // transformed into a proxy err error response.
+            bail!("did not receive response from render");
         }
     };
 
-    let body = match operation.recv().await? {
-        RenderProxyIncomingMessage::Body { data: body } => body,
-        RenderProxyIncomingMessage::Error(error) => {
-            bail!(
-                trace_stack(
-                    error,
-                    intermediate_asset,
-                    intermediate_output_path,
-                    project_dir
-                )
-                .await?
-            )
-        }
-        _ => {
-            bail!("unexpected response from the Node.js process while reading response body")
-        }
+    let RenderItem::Headers(data) = first else {
+        bail!("did not receive headers from render");
     };
 
-    Ok(ProxyResult {
-        status,
-        headers,
-        body: body.into(),
-    })
+    let body = Body::from_stream(stream.map(|item| match item {
+        Ok(RenderItem::BodyChunk(b)) => Ok(b),
+        Ok(v) => Err(BodyError::new(format!("unexpected render item: {:#?}", v))),
+        Err(e) => Err(BodyError::new(format!(
+            "error streaming proxied contents: {}",
+            e
```

Comment:
Use `.context()` instead of format!
We want to avoid converting to string

Reply:
I agree, but I'm deferring to a follow up PR.